### ISSUE TITLE
Immutable Attribute Modification Excepion

### ISSUE THREAD

**Issue by mike1372** [chef/chef#7366]
## Description
I am currently testing one of our client's cookbooks with Chef 14 and have noted the below error. This error does not occur with Chef client version 13.9.4 which is what has been previously used, note no other changes have been made to the cookbook.

## Chef Version
14.2.0

## Platform Version
Windows Server 2012 R2

## Replication Case
I am compiling the following resource which appears to be what is causing the issue:

windows_feature 'install .Net Framework versions: 3.5, 3.0 & 2.0' do
  action :install
  feature_name 'NetFx3'
  all true
  source windows_sxs_path
end

## Client Output

```
* windows_feature[install .Net Framework versions: 3.5, 3.0 & 2.0] action install
           * windows_feature_dism[install .Net Framework versions: 3.5, 3.0 & 2.0] action install
             
             ================================================================================
             Error executing action `install` on resource 'windows_feature_dism[install .Net Framework versions: 3.5, 3.0 & 2.0]'
             ================================================================================
             
             Chef::Exceptions::ImmutableAttributeModification
             ------------------------------------------------
             Node attributes are read-only when you do not specify which precedence level to set. To set an attribute use code like `node.default["key"] = "value"'
             
             Resource Declaration:
             ---------------------
             # In C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource/windows_feature.rb
             
       83:           declare_resource(subresource, new_resource.name) do
       84:             action desired_action
       85:             feature_name new_resource.feature_name
       86:             source new_resource.source if new_resource.source
       87:             all new_resource.all
       88:             timeout new_resource.timeout
       89:             management_tools new_resource.management_tools if subresource == :windows_feature_powershell
       90:           end
       91:         end
             
             Compiled Resource:
             ------------------
             # Declared in C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource/windows_feature.rb:83:in `run_default_subresource'
             
             windows_feature_dism("install .Net Framework versions: 3.5, 3.0 & 2.0") do
        action [:install]
        default_guard_interpreter :default
        declared_type :windows_feature_dism
        cookbook_name "xxxxxxxxxx"
        feature_name ["netfx3"]
        all true
        timeout 600
        source "C:\\Users\\ADMINI~1\\AppData\\Local\\Temp\\kitchen\\cache/windows_source_media/sxs"
             end
             
             System Info:
             ------------
             chef_version=14.2.0
             platform=windows
             platform_version=6.3.9600
             ruby=ruby 2.5.1p57 (2018-03-29 revision 63029) [x64-mingw32]
             program_name=C:/opscode/chef/bin/chef-client
             executable=C:/opscode/chef/bin/chef-client
             
           
           ================================================================================
           Error executing action `install` on resource 'windows_feature[install .Net Framework versions: 3.5, 3.0 & 2.0]'
           ================================================================================
           
           Chef::Exceptions::ImmutableAttributeModification
           ------------------------------------------------
           windows_feature_dism[install .Net Framework versions: 3.5, 3.0 & 2.0] (C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource/windows_feature.rb line 83) had an error: Chef::Exceptions::ImmutableAttributeModification: Node attributes are read-only when you do not specify which precedence level to set. To set an attribute use code like `node.default["key"] = "value"'
           
           Resource Declaration:
           ---------------------
           # In C:/Users/ADMINI~1/AppData/Local/Temp/kitchen/cache/cookbooks/xxxxxxxxxx/recipes/w2k12_dotnet3x.rb
           
            26: windows_feature 'install .Net Framework versions: 3.5, 3.0 & 2.0' do
            27:   action :install
            28:   feature_name 'NetFx3' # Case-sensitive now in version 3.* due to ohai lookup. Ensure case matches: 'dism /online /get-features' output.
            29:   # feature_name 'NET-Framework-Core'
            30:   all true
            31:   source windows_sxs_path
            32:   # install_method :windows_feature_powershell # Note exiting bug: https://github.com/chef-cookbooks/windows/issues/517
            33: end
           
           Compiled Resource:
           ------------------
           # Declared in C:/Users/ADMINI~1/AppData/Local/Temp/kitchen/cache/cookbooks/xxxxxxxxxx/recipes/w2k12_dotnet3x.rb:26:in `from_file'
           
           windows_feature("install .Net Framework versions: 3.5, 3.0 & 2.0") do
             action [:install]
             default_guard_interpreter :default
             declared_type :windows_feature
             cookbook_name "xxxxxxxxxx"
             recipe_name "w2k12_dotnet3x"
             all true
             feature_name "NetFx3"
             source "C:\\Users\\ADMINI~1\\AppData\\Local\\Temp\\kitchen\\cache/windows_source_media/sxs"
           end
           
           System Info:
           ------------
           chef_version=14.2.0
           platform=windows
           platform_version=6.3.9600
           ruby=ruby 2.5.1p57 (2018-03-29 revision 63029) [x64-mingw32]
           program_name=C:/opscode/chef/bin/chef-client
           executable=C:/opscode/chef/bin/chef-client
```

## Stacktrace
```
Generated at 2018-06-13 15:07:26 +1000
Chef::Exceptions::ImmutableAttributeModification: windows_feature[install .Net Framework versions: 3.5, 3.0 & 2.0] (xxxxxxxxxx::w2k12_dotnet3x line 26) had an error: Chef::Exceptions::ImmutableAttributeModification: windows_feature_dism[install .Net Framework versions: 3.5, 3.0 & 2.0] (C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource/windows_feature.rb line 83) had an error: Chef::Exceptions::ImmutableAttributeModification: Node attributes are read-only when you do not specify which precedence level to set. To set an attribute use code like `node.default["key"] = "value"'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/node/mixin/immutablize_array.rb:169:in `block (2 levels) in <module:ImmutablizeArray>'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource/windows_feature_dism.rb:122:in `features_to_install'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource/windows_feature_dism.rb:62:in `block in <class:WindowsFeatureDism>'
(eval):2:in `block in action_install'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/provider.rb:236:in `instance_eval'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/provider.rb:236:in `compile_and_converge_action'
(eval):2:in `action_install'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/provider.rb:182:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource.rb:577:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:70:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:98:in `block (2 levels) in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:98:in `each'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:98:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/resource_list.rb:94:in `block in execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/resource_list.rb:92:in `execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:97:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/provider.rb:237:in `compile_and_converge_action'
(eval):2:in `action_install'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/provider.rb:182:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource.rb:577:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:70:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:98:in `block (2 levels) in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:98:in `each'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:98:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/resource_list.rb:94:in `block in execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/resource_list.rb:92:in `execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:97:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/client.rb:717:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/client.rb:712:in `catch'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/client.rb:712:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/client.rb:751:in `converge_and_save'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/client.rb:283:in `run'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/application.rb:303:in `run_with_graceful_exit_option'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/application.rb:279:in `block in run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/local_mode.rb:44:in `with_server_connectivity'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/application.rb:261:in `run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/application/client.rb:441:in `run_application'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/application.rb:66:in `run'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/bin/chef-client:26:in `<top (required)>'
C:/opscode/chef/bin/chef-client:75:in `load'
C:/opscode/chef/bin/chef-client:75:in `<main>'

>>>> Caused by Chef::Exceptions::ImmutableAttributeModification: Node attributes are read-only when you do not specify which precedence level to set. To set an attribute use code like `node.default["key"] = "value"'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/node/mixin/immutablize_array.rb:169:in `block (2 levels) in <module:ImmutablizeArray>'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource/windows_feature_dism.rb:122:in `features_to_install'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource/windows_feature_dism.rb:62:in `block in <class:WindowsFeatureDism>'
(eval):2:in `block in action_install'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/provider.rb:236:in `instance_eval'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/provider.rb:236:in `compile_and_converge_action'
(eval):2:in `action_install'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/provider.rb:182:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource.rb:577:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:70:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:98:in `block (2 levels) in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:98:in `each'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:98:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/resource_list.rb:94:in `block in execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/resource_list.rb:92:in `execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:97:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/provider.rb:237:in `compile_and_converge_action'
(eval):2:in `action_install'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/provider.rb:182:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource.rb:577:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:70:in `run_action'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:98:in `block (2 levels) in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:98:in `each'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:98:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/resource_list.rb:94:in `block in execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/resource_collection/resource_list.rb:92:in `execute_each_resource'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/runner.rb:97:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/client.rb:717:in `block in converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/client.rb:712:in `catch'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/client.rb:712:in `converge'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/client.rb:751:in `converge_and_save'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/client.rb:283:in `run'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/application.rb:303:in `run_with_graceful_exit_option'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/application.rb:279:in `block in run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/local_mode.rb:44:in `with_server_connectivity'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/application.rb:261:in `run_chef_client'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/application/client.rb:441:in `run_application'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/lib/chef/application.rb:66:in `run'
C:/opscode/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-14.2.0-universal-mingw32/bin/chef-client:26:in `<top (required)>'
C:/opscode/chef/bin/chef-client:75:in `load'
C:/opscode/chef/bin/chef-client:75:in `<main>'
```

**Comment by lamont-granquist** [comment]
that is from this line:

https://github.com/chef/chef/blob/v14.2.0/lib/chef/resource/windows_feature_dism.rb#L122

that is using `#concat` to write to the parent level node attributes, which is correctly throwing the exception.

**Comment by tas50** [comment]
@mike1372 Thanks for the report. I fixed this in #7370
